### PR TITLE
[SPARK-50426][PYTHON] Avoid static Python data source lookup when using builtin or Java data sources

### PR DESCRIPTION
**Pull request by allisonwang-db**
### What changes were proposed in this pull request?

This PR updates the data source lookup logic to avoid triggering expensive Python data source lookups when the data source is a built-in or Java data source.

### Why are the changes needed?

Python data source lookups can be expensive, and they are often unnecessary when users only use built-in or Java data sources. The primary benefit of the lookup is to display a better error message when a Java data source and a Python data source share the same name, which can be rare. In these cases, we can simply prefer loading Java data sources instead of throwing exceptions.

### Does this PR introduce _any_ user-facing change?

No.

### How was this patch tested?

New UT.

### Was this patch authored or co-authored using generative AI tooling?

No

### CODE REVIEW

**Comment by allisonwang-db** [comment]
cc @HyukjinKwon @cloud-fan this is an alternative (and better) approach to https://github.com/apache/spark/pull/48888 to avoid necessary Python data source lookup. And it eliminates the need for users to change the static conf.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSource.scala [hunk 1]
```diff
@@ -682,10 +682,6 @@ object DataSource extends Logging {
                 throw e
               }
           }
-        case _ :: Nil if isUserDefinedDataSource =>
-          // There was DSv1 or DSv2 loaded, but the same name source was found
-          // in user defined data source.
-          throw QueryCompilationErrors.foundMultipleDataSources(provider)
```
Comment (HyukjinKwon) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSource.scala:688]: I think we should still throw an exception if one is found with a short name, but another one is found in Python Data Source.

Comment (HyukjinKwon) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSource.scala:688]: Can we at least document that Scala/JVM short names will be used when the short names are conflicted with Python Data Source?

Comment (HyukjinKwon) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSource.scala:688]: SG

Comment (allisonwang-db) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSource.scala:688]: Sg! Added usage notes in the python data source user guide.

**Review by HyukjinKwon — APPROVED** [review]

File: python/docs/source/user_guide/sql/python_data_source.rst [hunk 2]
```diff
@@ -516,3 +516,8 @@ The following example demonstrates how to implement a basic Data Source using Ar
     df = spark.read.format("arrowbatch").load()
 
     df.show()
+
+Usage Notes
+-----------
+
+- During data source resolution, built-in and Java data sources take precedence over Python data sources with the same name; to explicitly use a Python data source, make sure its name does not conflict with any other loaded Java data sources.
```
Comment (HyukjinKwon) [inline python/docs/source/user_guide/sql/python_data_source.rst:523]:
```suggestion
- During Data Source resolution, built-in and Scala/Java Data Sources take precedence over Python Data Sources with the same name; to explicitly use a Python Data Source, make sure its name does not conflict with the other Data Sources.
```

File: python/docs/source/user_guide/sql/python_data_source.rst [hunk 3]
```diff
@@ -516,3 +516,8 @@ The following example demonstrates how to implement a basic Data Source using Ar
     df = spark.read.format("arrowbatch").load()
 
     df.show()
+
+Usage Notes
+-----------
+
+- During Data Source resolution, built-in and Scala/Java Data Sources take precedence over Python Data Sources with the same name; to explicitly use a Python Data Source, make sure its name does not conflict with the other Data Sources.
```
Comment (HyukjinKwon) [inline python/docs/source/user_guide/sql/python_data_source.rst:523]: BTW, I think we should deal with the case when there are multiple Python Data Sources having the same names

**Comment by HyukjinKwon** [comment]
Merged to master.

**Pull request closed without merging**